presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DbManagerSpecProvider.java

```diff
@@ -0,0 +1,148 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.resourceGroups.db;
+
+import com.facebook.presto.resourceGroups.ManagerSpec;
+import com.facebook.presto.resourceGroups.ResourceGroupIdTemplate;
+import com.facebook.presto.resourceGroups.ResourceGroupSelector;
+import com.facebook.presto.resourceGroups.ResourceGroupSpec;
+import com.facebook.presto.resourceGroups.SelectorSpec;
+import com.facebook.presto.resourceGroups.reloading.ManagerSpecProvider;
+import com.facebook.presto.resourceGroups.reloading.ReloadingResourceGroupConfig;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Inject;
+import io.airlift.units.Duration;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class DbManagerSpecProvider
+        implements ManagerSpecProvider
+{
+    private final ResourceGroupsDao resourceGroupsDao;
+    private final String environment;
+
+    @Inject
+    public DbManagerSpecProvider(ResourceGroupsDao dao, @ForEnvironment String environment, ReloadingResourceGroupConfig config)
+    {
+        this.resourceGroupsDao = requireNonNull(dao, "dao is null");
+        this.environment = requireNonNull(environment, "environment is null");
+        this.resourceGroupsDao.createResourceGroupsGlobalPropertiesTable();
+        this.resourceGroupsDao.createResourceGroupsTable();
+        this.resourceGroupsDao.createSelectorsTable();
+        if (config.getExactMatchSelectorEnabled()) {
+            this.resourceGroupsDao.createExactMatchSelectorsTable();
+        }
+    }
+
+    @Override
+    public synchronized ManagerSpec getManagerSpec()
+    {
+        Set<Long> rootGroupIds = new HashSet<>();
+        Map<Long, ResourceGroupSpec> resourceGroupSpecMap = new HashMap<>();
+        Map<Long, ResourceGroupIdTemplate> resourceGroupIdTemplateMap = new HashMap<>();
+        Map<Long, ResourceGroupSpecBuilder> recordMap = new HashMap<>();
+        Map<Long, Set<Long>> subGroupIdsToBuild = new HashMap<>();
+        populateFromDbHelper(recordMap, rootGroupIds, resourceGroupIdTemplateMap, subGroupIdsToBuild);
+
+        // Build up resource group specs from leaf to root
+        for (LinkedList<Long> queue = new LinkedList<>(rootGroupIds); !queue.isEmpty(); ) {
+            Long id = queue.pollFirst();
+            resourceGroupIdTemplateMap.computeIfAbsent(id, k -> {
+                ResourceGroupSpecBuilder builder = recordMap.get(id);
+                return ResourceGroupIdTemplate.forSubGroupNamed(
+                        resourceGroupIdTemplateMap.get(builder.getParentId().get()),
```

Comment:
for root nodes `builder.getParentId().get()` will throw a NoSuchElementException . Can you help me understand how this is safe.

Reply:
it looks like getParentId() is an Optional<Long>, so i think get() won't throw an error because you're passing Optional.empty() in, not null

Reply:
Optional.empty().get() will throw a NoSuchElementException

can we move lines 72 to 76 inside populateFromDbHelper

Reply:
I see this is safe (but pretty sketchy) because in populateFromDbHelper, we put all the rootGroupIds in the template map, so the computeIfAbsent() call will only compute entries for non-root groups.  

@yyujiafb, I know this code existed and you just moved it, so I think it's fine to leave as is for this PR. But, it would be great to think about if there's a way to refactor this code to have fewer assumptions and interdependencies like this.

Reply:
just unresolving so you see the above comment :)